File: js/src/payload.ts
/** Whole-document compression presets, independent of numeric encodings. */
import { zlibCompress, zlibDecompress } from "./zlibp.js"

export type PayloadCompression = "raw" | "zlib" | "brotli" | "auto" | "r" | "z" | "b"
export interface BrotliBackend {
  compress(data: Uint8Array): Uint8Array
  decompress(data: Uint8Array, cap: number): Uint8Array
}
let brotli: BrotliBackend | undefined
export function registerBrotliBackend(backend: BrotliBackend): void { brotli = backend }
function requireBrotli(): BrotliBackend {
  if (!brotli) throw Error("Brotli support is unavailable. Call installBrotli() from '@spectrl-ms/spectrl/brotli'.")
  return brotli
}
const names: Record<string, string> = { raw: "r", zlib: "z", brotli: "b", r: "r", z: "z", b: "b" }
export function compressPayload(raw: Uint8Array, compression: PayloadCompression, cap: number): [string, Uint8Array] {
  if (raw.length > cap) throw Error("CBOR payload exceeds the size limit")
  if (compression === "auto") {
    const modes: PayloadCompression[] = ["z", "r"]
    if (brotli) modes.push("b")
    let best: [string, Uint8Array] | undefined
    for (const mode of modes) {
      const candidate: [string, Uint8Array] = [mode, pack(raw, mode)]
      if (candidate[1].length > cap) continue
      if (!best || candidate[1].length < best[1].length) best = candidate
    }
    return best!
  }
  const mode = Object.prototype.hasOwnProperty.call(names, compression) ? names[compression] : undefined
  if (!mode) throw Error("payload compression must be raw, zlib, brotli, or auto")
  const packed = pack(raw, mode)
  if (packed.length > cap) throw Error("compressed payload exceeds the size limit")
  return [mode, packed]
}
function pack(raw: Uint8Array, mode: string): Uint8Array {
  return mode === "r" ? raw : mode === "z" ? zlibCompress(raw, 6)
    : requireBrotli().compress(raw)
}
export function decompressPayload(packed: Uint8Array, mode: string, cap: number): Uint8Array {
  const raw = mode === "r" ? packed : mode === "z" ? zlibDecompress(packed, cap)
    : requireBrotli().decompress(packed, cap)
  if (raw.length > cap) throw Error("expanded CBOR payload exceeds the size limit")
  return raw
}

File: src/spectrl/payload.py
"""Whole-document compression presets with bounded decompression."""

from __future__ import annotations

import importlib.util
import zlib

from .codecs._zlibutil import bounded_decompress

NAMES = {"raw": "r", "zlib": "z", "brotli": "b", "r": "r", "z": "z", "b": "b"}


def _brotli():
    try:
        import brotli
    except ImportError as exc:
        raise ValueError("Brotli support is unavailable. Install spectrl[brotli].") from exc
    return brotli


def compress_payload(raw: bytes, compression: str, cap: int) -> tuple[str, bytes]:
    if len(raw) > cap:
        raise ValueError(f"CBOR payload exceeds {cap} bytes")
    if compression == "auto":
        modes = ["z", "r"]
        if importlib.util.find_spec("brotli") is not None:
            modes.append("b")
        candidates = [(mode, _pack(raw, mode)) for mode in modes]
        candidates = [item for item in candidates if len(item[1]) <= cap]
        return min(candidates, key=lambda item: len(item[1]))
    try:
        mode = NAMES[compression]
    except (KeyError, TypeError) as exc:
        raise ValueError("payload compression must be raw, zlib, brotli, or auto") from exc
    packed = _pack(raw, mode)
    if len(packed) > cap:
        raise ValueError(f"compressed payload exceeds {cap} bytes")
    return mode, packed


def _pack(raw, mode):
    if mode == "r":
        return raw
    if mode == "z":
        return zlib.compress(raw, 6)
    return _brotli().compress(raw, quality=5)


def decompress_payload(packed: bytes, mode: str, cap: int) -> bytes:
    if mode == "r":
        return packed
    if mode == "z":
        return bounded_decompress(packed, cap)
    if mode != "b":
        raise ValueError(f"unsupported payload mode: {mode!r}")
    decoder = _brotli().Decompressor()
    chunks = []
    count = 0
    data = packed
    while True:
        chunk = decoder.process(data, output_buffer_limit=min(65536, cap - count + 1))
        count += len(chunk)
        if count > cap:
            raise ValueError(f"brotli output exceeds the {cap}-byte limit")
        chunks.append(chunk)
        if decoder.is_finished():
            return b"".join(chunks)
        if not chunk:
            raise ValueError("truncated brotli payload")
        data = b""
